### DISCUSSION TITLE
Is it possible to use storage buffers for web apps?

### DISCUSSION THREAD

**Discussion by HansWestman74** [gfx-rs/wgpu#4124] [Q&A] [answered]
Hi,

I'm wondering if there is anyway I can use storage buffers for a web application and wgpu code compiled to wasm.

I'm sitting on a Macbook Pro M2. My app works fine when I compile to native and metal as backend. But when I compile to wasm and run the app in a browser I get the following error message in the console:

"In Device::create_bind_group_layout
      note: label = `bind_group_layout`
    Too many bindings of type StorageBuffers in Stage ShaderStages(FRAGMENT), limit is 0, count was 4"
    
The backend has defaulted to Gl.

Is there anyway at this time I can use storage buffers in a web app? Or do we only have native support for storage buffers at this time?

Many thanks for any help!

**Comment by cwfitzgerald** [comment]
WebGL2 doesn't suppe ort storage buffers at all. If you want to use storage buffers on the web, you need to use WebGPU. WebGPU is only supported onh chrome (on windows, mac, and chromeos) currently. To do so, you need to remove the webgl feature from thw wgpu crate. we will then run on top of WebGPU, which does support storage buffers.

**Reply by HansWestman74** [reply]
Thanks for the answer! Can you expand a bit more on how to do this:

"remove the webgl feature from thw wgpu crate"

Is this something I can do from .toml file? Or is it more involved?

**Comment by HansWestman74** [comment]
When I remove the "webgl" feature I get a lot of errors such as 

"
use web_sys::GpuTextureFormat as tf;
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `GpuTextureFormat` in the root"
    
    Do I need to enable some other feature to get rid of these errors? What's the best way to proceed here?

**Reply by daxpedda** [reply, marked as answer]
You need to activate a `cfg` flag for `web-sys` to export all these types.
E.g. `RUSTFLAGS=--cfg=web_sys_unstable_apis`.

**Comment by HansWestman74** [comment]
This worked! Many thanks for your help and support.
